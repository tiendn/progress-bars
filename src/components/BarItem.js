import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';

const BarItem = ({ active, limit, value }) => {
  const percent = Math.floor((value / limit) * 100);
  const width = percent < 100 ? percent : 100;

  const backgroundClassName = classNames(
    'barItemHighlight',
    value > limit ? 'barItemLimit' : '',
    active ? 'barItemActive' : 'barItemInactive',
  );

  const textValueClassName = classNames(
    'barItemTextValue',
    width < 55 ? 'barItemTextValueNegative' : '',
  );

  return (
    <div className="barItem">
      <div
        data-testid="barItemHightlight"
        className={backgroundClassName}
        style={{ width: `${width}%` }}
      />
      <span data-testid="barItemValue" className={textValueClassName}>
        {percent}%
      </span>
    </div>
  );
};

BarItem.propTypes = {
  active: PropTypes.bool.isRequired,
  limit: PropTypes.number.isRequired,
  value: PropTypes.number.isRequired,
};

export default BarItem;
