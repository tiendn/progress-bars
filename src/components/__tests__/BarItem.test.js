import React from 'react';
import ReactDOM from 'react-dom';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import BarItem from '../BarItem';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<BarItem value={20} limit={100} active />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('renders matching snapshot', () => {
  const { baseElement } = render(<BarItem value={120} limit={100} active />);
  expect(baseElement).toMatchSnapshot();
});

it('render right value', async () => {
  const value = 120;
  const { findByTestId } = render(<BarItem value={value} limit={100} active />);
  const textValue = await findByTestId('barItemValue');

  expect(textValue.innerHTML).toBe(`${value}%`);
});

it('render with class "barItemLimit" when the value over limit', async () => {
  const { findByTestId } = render(<BarItem value={120} limit={100} active />);
  const divHighlight = await findByTestId('barItemHightlight');

  expect(divHighlight).toHaveClass('barItemLimit');
});
