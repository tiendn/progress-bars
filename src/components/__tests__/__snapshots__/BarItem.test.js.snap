// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders matching snapshot 1`] = `
<body>
  <div>
    <div
      class="barItem"
    >
      <div
        class="barItemHighlight barItemLimit barItemActive"
        data-testid="barItemHightlight"
        style="width: 100%;"
      />
      <span
        class="barItemTextValue"
        data-testid="barItemValue"
      >
        120
        %
      </span>
    </div>
  </div>
</body>
`;
